src/interfaces/MetalsInputBox.ts

```diff
@@ -0,0 +1,137 @@
+import {
+  RequestType,
+  Range,
+  MarkupContent,
+} from "vscode-languageserver-protocol";
+
+/**
+ * The Metals input box request is sent from the server to the client to
+ * let the user provide a string value for a given prompt. Unlike
+ * `window/showMessageRequest`, the `metals/inputBox` request allows the user
+ * to provide a custom response instead of picking a pre-selected value.
+ *
+ * - https://scalameta.org/metals/docs/editors/new-editor.html#metalsinputbox
+ */
+export namespace MetalsInputBox {
+  export const type = new RequestType<
+    InputBoxOptions,
+    MetalsInputBoxResult,
+    void,
+    void
+  >("metals/inputBox");
+}
+
+export interface MetalsInputBoxResult {
+  value?: string;
+  cancelled?: boolean;
+}
+
+/**
+ * The below is ported directly from VS Code:
```

Comment:
This is a bit unfortunate, but since this is tied to the vs code api, I don't see a way around it. This won't really help the vs code extension much, but I have all of this duplicated in coc-metals. I was opposed to moving this over at first, but then realized if this api changes, then the Metals implementation needs to change anyways. So, I figured this was ok to port over. I was also going to do the same thing with decorations, but that goes _way_ deeper than this, and I've have to port over like 3 times this amount of stuff, which I don't think is a good idea. That's way the decoration protocol isn't found in here.

Reply:
I see the problem. One way to work around the issue is to define the protocol interface here in an "agnostic way" and then to have specific implementations of the protocol in the clients.

Methods in metals-languageclient that need to use an input box can require the implementation from the client in a "manual typeclass" fashion.

Does it make any sense?

Reply:
The idea makes sense yes. How to implement this in TS alludes me a bit. Could you give me an example?

Reply:
Nothing fancy at all:

```ts
// in metals-languageclient

// the "typeclass" definition
interface MetalsInputBox {
  // or something less lazy than this :-P
  function askStuffToUser(message: string): Promise<string>;
}

// some method require the typeclass instance
function someMethodThatAsksThings(inputBox: MetalsInputBox): (message: string) => Promise<unknown> {
  return message => inputBox.askStuffToUser(message).then(...)
}
```

```ts
// in metals-vscode

// instance definition
const inputBox: MetalsInputBox = {
  askStuffToUser: message => {
    // implementation using VSCode API
  }
}

// "bind" the method to the typeclass (optional, but may be useful if invoked a lot)
const askThings = someMethodThatAsksThings(inputBox)

// ...

// use the bound method
askThings("hey, what's your name?").then(...)
```

Reply:
🤔 I guess I'm struggling with seeing the benifit of this, and the more I think of it, the more I'm actually thinking that porting this into here does make sense. Mainly, I don't think this extra layer of abstraction will gain us anything. For example, this is only used in the following two places right:

In VS Code
```ts
    client.onRequest(MetalsInputBox.type, (options, requestToken) => {
      return window.showInputBox(options, requestToken).then((result) => {
        if (result === undefined) {
          return { cancelled: true };
        } else {
          return { value: result };
        }
      });
    });
```

In coc-metals
```
    client.onRequest(MetalsInputBox.type, (options: InputBoxOptions) => {
      return workspace.callAsync<string>("input", [
        `${options.prompt}: `,
        options.value ? options.value : "",
      ]).then((response) => {
        if (response.trim() === "") {
          return { cancelled: true };
        } else {
          return { value: response };
        }
      })
    });
```

So we would basically create, bind, and use the instance once. I could see this being useful if we used it all over the place, but we don't. Also, if we would do the typeclass approach, I'd still need to define an interface in coc-metals for `InputBoxOptions` which is used.  I sort of think this is just a result of us exactly using `InputBoxOptions` from the VS Code api, and it actually should be part of the shared library if it's part of the protocol.

Thoughts?

Reply:
Basically just adding the following to the namespace:

```ts
  export const handleInput = function (result: string | undefined): MetalsInputBoxResult {
    if (result === undefined || result.trim() === "") {
      return {cancelled: true};
    } else {
      return {value: result};
    }
  }
```

Which would then allow for:

VS Code
```ts
    client.onRequest(MetalsInputBox.type, (options, requestToken) => {
      return window
        .showInputBox(options, requestToken)
        .then(MetalsInputBox.handleInput);
    });
```

coc-metals

```ts
InputBox.type, (options: InputBoxOptions) => {
      return workspace
        .callAsync<string>("input", [
          `${options.prompt}: `,
          options.value ? options.value : "",
        ])
        .then(MetalsInputBox.handleInput);
    });
```